Allow tabs to wrap to multi-line

SneakyFish5 opened the pull request:
<!-- Thank you for submitting a Pull Request. Please: 
* Read our Pull Request guidelines:
  https://github.com/Microsoft/vscode/wiki/How-to-Contribute#pull-requests.
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `master` branch.
* Include a description of the proposed changes and how to test them. 
-->

This PR fixes https://github.com/microsoft/vscode/issues/70413

![M5b7NlCmUZ](https://user-images.githubusercontent.com/32284796/92657121-dbe04100-f2b9-11ea-91ad-affa340a50d8.gif)

bpasero reviewed (changes requested):
@SneakyFish5 kudos, this is a very elegant solution given it only requires CSS changes 👍 

Some feedback after a quick review:
* can we expose the setting from [`IEditorPartConfiguration`](https://github.com/microsoft/vscode/blob/4e1109bdefaa12ef5d5c6809522aa748fb0b468e/src/vs/workbench/common/editor.ts#L1220-L1220)? This should make it much easier to read the setting and react to changes (e.g. from https://github.com/microsoft/vscode/blob/82d953a731d3acf530f2ee898536b58ac3d00e05/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L124-L124)
* the color introduced for the border needs to be themable (i.e. cannot be hardcoded in the CSS) and needs some testing together with the existing tab border colors that we have, I am not sure it will work in all cases
* when tabs wrap they do not align nicely below the editor action bar (see screenshot below), can we possibly improve this?
* there seems to be a bug when trying to drop a tab after the last tab, the drop feedback spans multiple rows (see video below)

![image](https://user-images.githubusercontent.com/900690/92924659-9cd8f980-f439-11ea-909c-afcc490151bd.png)

![recording (1)](https://user-images.githubusercontent.com/900690/92924765-c1cd6c80-f439-11ea-9b5c-b0c66d475803.gif)

bpasero commented:
@SneakyFish5 

> I think a neat solution for both of these problems is provided by @fraigo in #70413 (comment), let me know what you think. Here's a visual representation of it included with all the other changes:

This looks very similar to `tabSizing: shrink` where tabs shrink to a fixed size when too many are opened. What I do not like about the presented solution is that tabs now consume the entire width even when available space is there to show them normally. Ideally only tabs at the far right end would align vertically by growing or shrinking, right? Not sure how to get that done in pure CSS though... maybe there is a flex property that can help?

fraigo commented:
@bpasero @SneakyFish5 One solution to make the last row use the space better is adding a 'hidden' flex space after the last tab (in this case, adding space at the end of the tab container. I think `33%` is a good value, could be changed to any other for the try the best fit.

```
.monaco-workbench .part.editor > .content .editor-group-container > .title.tabs> .tabs-and-actions-container.multi-line-tabs-container > .monaco-scrollable-element > .tabs-container::after {
	content: "";
	flex: 1 0 33%;
}

```

![multi_tabls_last](https://user-images.githubusercontent.com/8151943/93124917-36451d00-f67f-11ea-9d3d-2dc1c248af62.gif)

SneakyFish5 commented:
After some digging as well I couldn't figure out how to make the tabs stay the same size and align properly at the very far end, aside from setting a `max-width` to an arbitrary percentage, which is a pretty hacky fix. I like the solution that @fraigo is presenting. When testing it out `flex: 1 0 0;` worked better for me and made a smoother transition to the next line when opening new tabs, would there be any problem with that?

fraigo commented:
yes @SneakyFish5, using `flex:1 0 0` works fine (even using `flex: 1 0 auto` should work in this case because there is no content to shrink), using the minimum spanning space.

SneakyFish5 commented:
Awesome, recorded a little snippet here:

![Mq6p1kONjL](https://user-images.githubusercontent.com/32284796/93159884-d8432480-f6d4-11ea-9a47-968f3b01ed45.gif)

@bpasero is that better?

bpasero commented:
@SneakyFish5 can you add this solution to this PR for me to play with? Easier then to give some feedback. I will also comment on the tab border then, thanks 👍

SneakyFish5 commented:
Will do 👍

bpasero commented:
@SneakyFish5 this works great when many tabs are opened and it wraps but is there a way to not have excessively large tabs when you only have few opened:

![image](https://user-images.githubusercontent.com/900690/93299454-60205000-f7f5-11ea-8f34-7d5eec8f89ab.png)

?

bpasero commented:
I pushed a few polish changes to your branch and thought about the use of border: I think conceptually it makes a lot of sense to use `tab.border` for this because according to the docs: `Border to separate tabs from each other` so it fits.

However, I think the logic of drawing this border should merge with the existing border we already have for active tabs:

https://github.com/microsoft/vscode/blob/9dc00c63efa2b9c8529299cc55d5be9a85fa07a7/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1150-L1156

This ensures we are using the same logic for rendering a border and more importantly allows to still have a different border color for the active tab compared to others.

SneakyFish5 commented:
> this works great when many tabs are opened and it wraps but is there a way to not have excessively large tabs when you only have few opened

That's the current issue I'm facing that I'm not sure how to solve. I've tried setting a `max-width`, but I've realized it doesn't work on smaller window sizes. If anyone has any better ideas, please chime in as I'm definitely not the best at CSS.

> However, I think the logic of drawing this border should merge with the existing border we already have for active tabs which also draws a border below tabs

Done, it now only draws a bottom-border for inactive tabs when multi-line tabs are enabled. Is that what you meant?

The commits got a little messy there (sorry about that), but I've fixed merge conflicts and changed it to use `classList.toggle` instead.

bpasero commented:
@SneakyFish5 one thing we could maybe try is to do the flex trick conditionally depending on wether the tabs wrap or not? E.g. when you open only a few tabs they would appear normally and only when they wrap we start the trick?

Code wise we know exactly how much space we have and how much space tabs consume around here:

https://github.com/microsoft/vscode/blob/90a9985d51d3dc1afeb734316e09867558d3156c/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1317-L1317

So we could toggle a similar class depending on how much space tabs consume in the tab container?

SneakyFish5 commented:
That's definitely possible, I've made it so that the multi-line tabs class only added if the "all tabs width" is greater than the "visible tabs width", and is removed if they are equal and the container height is equal to `EDITOR_TITLE_HEIGHT`.

![FRIxM5PYVj](https://user-images.githubusercontent.com/32284796/93385921-8a413480-f82c-11ea-91ce-06dcd490cc33.gif)

I agree this probably isn't the most optimal situation and that it would be preferable to have the tabs be somewhat the same size and also expand to fill the container at the end of the row, but I'm not sure how to accomplish that.

bpasero commented on src/vs/workbench/browser/parts/editor/editorDropTarget.ts:
```diff
@@ -501,6 +502,10 @@ class DropOverlay extends Themable {
 
 	private getOverlayOffsetHeight(): number {
 		if (!this.groupView.isEmpty && this.accessor.partOptions.showTabs) {
+			const tablist = document.querySelector(DropOverlay.TABS_SELECTOR) as HTMLElement;
```
This will not work properly once you have more than one editor group opened because then you will have more than one `div.tabs-container`. 

I suggest an alternate solution where the `accessor` provides a method to ask for the height of the editor title area. The `accessor` in this case is the `EditorPart` which can delegate this to the group in question and then the group can forward this to the title control. And I would only use the `offsetHeight` property if mutli-line tabs are enabled.

SneakyFish5 replied:
Done, please let me know if I've done this incorrectly.

bpasero commented on src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:
```diff
@@ -517,15 +529,17 @@ export class TabsTitleControl extends TitleControl {
-		if (oldOptions.titleScrollbarSizing !== newOptions.titleScrollbarSizing) {
-			this.updateTabsScrollbarSizing();
-		}
+	private updateMultiLineTabs(): void {
```
There now seems to be a weird condition under which we seem to be adding the `multi-line` class when settings change but we conditionally add and remove it in the `layout` call. Shouldn't we exclusively be setting this class only from the layout call given that this is the only place where we know that we need it when the size is little?

I would say we let the `layout` method take care and make sure it is called when this setting changes.

SneakyFish5 replied:
Ah yes, I forgot about that when moving it to the `layout` method. I've removed it now, thanks.

bpasero commented on src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:
```diff
@@ -1174,6 +1188,10 @@ export class TabsTitleControl extends TitleControl {
 			tabContainer.style.backgroundColor = this.getColor(isGroupActive ? TAB_INACTIVE_BACKGROUND : TAB_UNFOCUSED_INACTIVE_BACKGROUND) || '';
 			tabContainer.style.boxShadow = '';
 
+			// bottom border when wrapping multi-line
```
My suggestion was to reuse the `tab-border-bottom` helper for this case too and not introduce yet another border. Currently the `tab-border-bottom` is reserved for active tabs because the border was only used for active tabs, but to ensure we get a consistent look, we should be using the same thing for the separator as well. Otherwise we end up with an active border that is not on the exact same line as the separator border.

SneakyFish5 replied:
Done, I misunderstood what you suggested earlier but hopefully it's correct now?

bpasero commented:
Another issue to look at is to somehow if possible prevent this overflow in wrapping that makes tab no longer accessible:

![multi-line-tabs-maximum](https://user-images.githubusercontent.com/900690/93493643-6fde8800-f90c-11ea-9682-9b75377ca9fb.png)

SneakyFish5 commented:
Could something like a vertical scroll bar be used for the overflowing tabs?

fraigo commented:
@SneakyFish5 Doing some checks, the scrollable container must limit vertical size. Not sure how much should be the limit (in this example, 30% of the viewport height). Also, it must handle the vertical scroll bar the same way horizontal scroll is managed in non-multi-line tabs (a nice-looking-thin scroll bar). 

src/vs/workbench/browser/parts/editor/media/tabstitlecontrol.css:49

```
.monaco-workbench .part.editor > .content .editor-group-container > .title.tabs > .tabs-and-actions-container.multi-line > .monaco-scrollable-element {
	height: auto !important;
	max-height: 30vh;
	overflow-y: auto !important;
}
```

![image](https://user-images.githubusercontent.com/8151943/93533668-46b3fc80-f8f8-11ea-9fa1-5cd00f4a1c19.png)

SneakyFish5 commented:
Thank you as always for your CSS work @fraigo, I'll to familiarize myself with the scrollbar and see if I can implement a vertical scrollbar. I'm assuming this would be done somewhere around here?:
https://github.com/microsoft/vscode/blob/3c6a1ad913d84802fa762f8849e17f35cdda7876/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L160

bpasero commented:
@SneakyFish5 thanks for the changes. I feel that the discussion about the border for tabs is something that we should look at separately because I think this matter is a bit more complex and I can also offer to add this after we merged this PR. I would like to focus exclusively on the multi-line wrapping in this PR, so I went ahead and removed the related changes https://github.com/microsoft/vscode/pull/106448/commits/e6f128801cb23cbc3e15f87bcab8dfaf4d5abe62 if you don't mind. 

**Things I did:**
* remove border handling and leave it for later
* cleanup use of editor group title (your change was exactly what I had in mind)
* cleanup a little bit the `doLayoutTabs` method to access `tabsContainer.offsetHeight` only when really needed (I am sensitive to accessing DOM properties that may trigger a layout in the browser)

Let's focus on the `doLayoutTabs` method: this is now THE place to decide if tabs wrap or not, which I think is good because there we know exactly how much width the tabs consume, so we can enable this conditionally. However I do not understand the logic you encoded there with regards to the `offsetHeight` and I also think that we now always set the `multi-line` class, even when not needed. Can we discuss this algorithm there?

As for the issue of having tabs exceed the height of the editor container: I would NOT add a scrollbar her but simply disable multi-line tabs IF the `offsetHeight` if the tabs exceeds a certain height. For example we can simply use the `DEFAULT_EDITOR_MIN_DIMENSIONS.height` as a driver and disable multi-line tabs if the editor height gets below this value?

SneakyFish5 commented:
Of course, I have no problem with any additions/removals that are needed. Please continue to do as you see fit, I'm also very keen on getting this merged so any thing necessary to move towards that goal is no issue with me.

> However I do not understand the logic you encoded there with regards to the offsetHeight and I also think that we now always set the multi-line class, even when not needed. Can we discuss this algorithm there?

Sure, my thinking was the following:

1. If `allTabsWidth` exceeds `visibleTabsContainerWidth`, then apply the `multi-line` class to wrap tabs if multi-line tabs have been enabled in settings
2. However, the flaw in the above is that once the tabs were wrapped,  `allTabsWidth` would once again be equal to `visibleTabsContainerWidth`, so the class would be immediately removed. So, if `allTabsWidth` and `visibleTabsContainerWidth` are equal, it also checks if `allTabsHeight` is equal to `EDITOR_TITLE_HEIGHT` (the default height). If it is, then the `multi-line` class is removed.

You mentioned that the class is set even when not needed, is there a specific example of this? If my logic is wrong somewhere I can attempt to revise it.

> For example we can simply use the DEFAULT_EDITOR_MIN_DIMENSIONS.height as a driver and disable multi-line tabs if the editor height gets below this value?

Can do, if possible could you give me a pointer for where to get the current editor height from? I can try to look for it/create a method for it myself if it doesn't already exist.

bpasero commented:
@SneakyFish5 ok I understand now and pushed https://github.com/microsoft/vscode/pull/106448/commits/318f2983d93a3c4b966a57340817bf7192ad5bf9 to clarify this. I am not able to see a case where the `multi-line` class is wrongly there, it seems to work fine.

As for the editor height: I think we need a similar thing as you added with `titleHeight` for `editorHeight`. Within the layout we should then add another check for the editor height becoming too small and then remove the `multi-line` class. I think the height of the editor can easily be computed because it is basically the available height of the entire editor group minus the title height.

SneakyFish5 commented:
Just pushed https://github.com/microsoft/vscode/pull/106448/commits/e6974fdfec48de8ae798eb340088f5fb7c5fde41 to fix a bug I noticed where the `editor-container` height wasn't being updated based on the changes in `titleHeight`. Let me know if you spot any problems with that.

As for the overflowing problem, I've run into some issues with that. I'm able to remove the `multi-line` class and get the tabs back to a single row when the `editorHeight` is below the minimum, however in doing so, the `editorHeight` becomes greater than the minimum again and repeats the process. https://github.com/microsoft/vscode/pull/106448/commits/3a3d7f0bf8aedf0d849e71154c1b6c7f2d9c24f7 is what I've gotten so far, and I'll continue to work on it, but I wanted to open up my progress to see if there were any possible solutions.

bpasero commented:
Yes you are right, you found a problem I have not thought about: normally the size of the title area was fixed (`35px` + breadcrumbs if visible) and as such we were able to do a good layout down to the editor, which is very important (it needs to be correct all the time).

With tabs now potentially wrapping, the size is no longer static and as such, a layout into the title can result in the title size changing. This is getting worse due to the fact that the title area layout is actually async:

https://github.com/microsoft/vscode/blob/09a78db5ef427ab945a2773239ab9fd6a9ef5e1c/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1239-L1239

I tried to come up with a solution today but could not come up with something simple. While the CSS solution is very elegant at first, we have have hit a blocker given we cannot easily provide the exact height down to the editor.

SneakyFish5 commented:
I'm slightly confused, I thought with the modifications to `getPreferredHeight` that this problem had been fixed and the title now dynamically, and correctly, returns it's height if multi-line tabs are enabled. Could you possibly elaborate on the issues that need fixing (aside from the overflowing tabs as I'm still not able to figure that out), so that I might be able to work on fixing them?

bpasero commented:
@SneakyFish5 I think there is a weird cycle now between `editorGroupView#layout` and `tabsTitleControl#layout`: 
* first we ask the tabs title control for the height using `offsetHeight` which can trigger a full browser layout btw
* then we ask to layout the tabs title control
* this potentially results in the title control to change its height depending on the logic inside
* but we still use the previous computed height for the editor
* we may end up using the wrong height for the editor

I think conceptually it should be:
* editorGroupView#layout which will trigger...
* ...title control layout which returns its height that it ended up using
* then editor control layout with that height

With this algorithm I would also think that you do not need to know the editor height, you can simply pass the overall height into the title control layout and then decide from within if the title should wrap or not.

Does that make sense?

As for the additional challenge that a `redraw` of tabs can result in the title height changing: I think this should not trigger a `relayout`, rather I would think that the tabs title control should have an event that is fired whenever the height changes and the editor group can register on that to trigger `redraw`. I think that would also help for the current logic that triggers `redraw` when breadcrumbs are enabled or disabled.

**PS:** I think it would be good if the CSS changes in tabsTitleControl.css could get some comments, I am having a hard time understanding the need for the properties, as well as the `important` and it would help for future maintenance if the logic is explained via comments.

SneakyFish5 commented:
Correct me if I'm wrong, but would the change be made here?:
https://github.com/microsoft/vscode/blob/c1905a3d139ea803c8f46ffc6535e9abdfcb1b60/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L1702

If I understood your concerns correctly, when the `titleHeight` is increased/decreased due to a new row of tabs being added/removed, the `editorHeight` doesn't properly update when `editorGroupView#layout` is called? When testing, I don't see that issue, am I missing something?:

![MoYeiJdP1Z](https://user-images.githubusercontent.com/32284796/93913380-a03a7380-fcca-11ea-8919-368eeabe22f3.gif)

>-  ...title control layout which returns its height that it ended up using
>-  then editor control layout with that height

I can definitely try to cleanup the logic and have `titleAreaControl#layout` return it's height and recalculate the `editorHeight` based upon that.

> With this algorithm I would also think that you do not need to know the editor height, you can simply pass the overall height into the title control layout and then decide from within if the title should wrap or not.

Again I probably misunderstand what's going on, but is this not already being done in the `editorGroupView#layout` that was linked above? The `editorHeight` property I created for the group view was an attempt at fixing the overflowing tabs issue, I can remove that for now since it's currently not working/unnecessary.

> I would think that the tabs title control should have an event that is fired whenever the height changes and the editor group can register on that to trigger redraw.

Sure, I can create an event that fires when the height of the `tabsContainer` changes from what it previously was. However, I don't think the editor group can call `redraw`, would `editorGroupView#relayout` instead be called, thereby replacing this line?:

https://github.com/microsoft/vscode/blob/c1905a3d139ea803c8f46ffc6535e9abdfcb1b60/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1023-L1027

> I think it would be good if the CSS changes in tabsTitleControl.css could get some comments

Done, sorry about that. Hopefully they're a bit more descriptive now, if it needs further improvements let me know. I removed the `!important` in both classes since I realized they're unnecessary, forgot why they were there in the first place. I also fixed `editorDropTarget` to make use of the new changes in `master`.

This got to be very big, did not mean for that to happen haha, I just want to run everything by you especially with the new changes to `master` that I needed to familiarize myself with.

swinder0161 commented:
Hi @SneakyFish5 @bpasero This looks great.

When can this pull request go into the master?

SneakyFish5 commented:
I'm sure @bpasero is very busy and will get around to it if/when he can. Nothing to do but be patient, on my end though I'll do what I can to get it merged 😄

bpasero commented:
Sorry for the silence, did not have a lot of time on this, will try to get back here in October.

WillcoWD commented:
Sorry if this is the wrong place to ask, but this is the only thing stopping me from swtiching to VSCode. 
I've cloned the Repo but can't see how to start VSCode on linux

SneakyFish5 commented:
Your feedback would be much appreciated @WillcoWD, instructions for building vscode can be found here: https://github.com/microsoft/vscode/wiki/How-to-Contribute

bpasero commented:
@SneakyFish5 

> Correct me if I'm wrong, but would the change be made here?:

Yes, the point where the layout for both editor and title widget starts is this:

https://github.com/SneakyFish5/vscode/blob/c1905a3d139ea803c8f46ffc6535e9abdfcb1b60/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L1702-L1702

As I already said, wouldn't this layout algorithm make more sense:
* editor group view layout gets called with the size it can consume
* editor title layout is called with that dimension and returns the actual width/height it consumes based on some rules
  * if multi line tabs are disabled, it just returns with `35px` (as today)
  * if multi-line tabs are enabled but the resulting editor size is below min width, it just returns with `35px` (see note below)
  * otherwise it returns with the exact height it consumed based on `offsetHeight`
* editor group uses the title height to pass on to the editor control
* layout is done

**Note:** maybe instead of the title control having hardcoded knowledge about the minimum size of the editor, it could rather accept the actual dimensions + the max dimensions it can consume. That way, the editor group view layout call can pass both the dimension and the max dimension into the title control. 

I suggest for now we ignore the other feedback I gave about `redraw` calling `relayout`, I think this needs additional thoughts how to get right. Let's first focus on the layout algorithm from editor group to editor title.

SneakyFish5 commented:
Ok so I've attempted to rework the layout algorithm based on your comments but I'm not sure if I've done it correctly.

I've made it so that `tabsTitleControl#layout` now returns the height using `getDimensions().height`, since it seems to already implement the rules you've outlined?
https://github.com/microsoft/vscode/blob/ac0a02482669a6b142d0d09f516f733fe2b906c5/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1276

Then, the new height returned by the title area layout call is then passed directly into `editorControl#layout`.
https://github.com/microsoft/vscode/blob/ac0a02482669a6b142d0d09f516f733fe2b906c5/src/vs/workbench/browser/parts/editor/editorGroupView.ts#L1706-L1710

I didn't quite understand what you meant by passing in 'max dimensions' into the title control, could you elaborate on that?

P.S. - I realized the `!important` on `.monaco-scrollable-element` was necessary otherwise multi-line tabs would mess up with breadcrumbs so I've readded it in.

bpasero commented on src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:
```diff
@@ -1253,6 +1272,8 @@ export class TabsTitleControl extends TitleControl {
 			});
 		}
+
+		return this.getDimensions().height;
```
As you can see from the code above, the layout is scheduled `async` to prevent perf issues. As such, you return a dimension here from a time before the layout is actually applied. This is one of the issues I talked about earlier, we cannot return the height until we actually applied the CSS classes. I fear we probably have to switch to a sync layout when multi-line tabs are enabled.

SneakyFish5 replied:
Ah true, I overlooked that. I'm not very familiar with how the delegation of the layout to a new animation frame works, so I may need some assistance on the implementation here.

bpasero commented on src/vs/workbench/browser/parts/editor/editorGroupView.ts:
```diff
@@ -1698,14 +1702,12 @@ export class EditorGroupView extends Themable implements IEditorGroupView {
-		this.editorContainer.style.height = `${editorHeight}px`;
+		// Use the initial height of the title area to layout
+		const initialTitleHeight = this.titleDimensions.height;
+		const titleHeight = this.titleAreaControl.layout(new Dimension(width, initialTitleHeight)) as number;
```
Can we not simply pass the full dimensions over that we get above instead of asking the title for its height? Together with the max dimensions for the title widget I thought along the lines of:
* one parameter is the dimension of the entire group and the title widget can use it up to...
* a second parameter defines a certain maximum of size the title widget can have

Based on these 2 parameters the title widget then decides wether it goes multi-line or not.

SneakyFish5 replied:
Ok I can try and start working on that, what would you suggest as the maximum size the title widget can have?

bpasero replied:
I would think that the maximum size is:
* `width`: the same as the entire width, aka there is no maximum
* `height`: the entire height minus the minimal height of the editor control

SneakyFish5 replied:
Ok I just pushed a very rough implementation that uses a `maxDimensions` parameter to try and remove multi-line tabs. There are problems that arise when testing, presumably due to scheduling the layout on the next frame. I'm sure it needs some fixing so let me know what you think.

bpasero commented on src/vs/workbench/browser/parts/editor/titleControl.ts:
```diff
@@ -407,7 +407,7 @@ export abstract class TitleControl extends Themable {
 	abstract updateStyles(): void;
 
-	abstract layout(dimension: Dimension): void;
+	abstract layout(dimension: Dimension): number | undefined;
```
For consistency I would suggest that `layout` returns both `height` and `width` (as a `Dimension`) and not just the one property. I would think that the `width` property can be used that is passed in because the title widget always gets as wide as what is given.

SneakyFish5 replied:
Done, it now returns a `Dimension`. It still has the problem of returning a `Dimension` before the layout call actually happens on the next scheduled frame, but once we figure out how to proceed with that, I'm sure it can be easily fit in.

bpasero reviewed (changes requested):
@SneakyFish5 tried to explain your questions in code. I realize this is not an easy ask around the layout, but I think this is also not an easy problem to solve, so I would not want to push a half-baked solution.

yashbhutwala reviewed (approved):
utACK

bpasero commented:
Yeah I think we need to drop the async layout code altogether if multi-line tabs are enabled. If multi-line tabs are enabled we cannot do a correct layout unless we layout sync and return the actual dimensions after wrapping. This is unfortunate but I see no  other solution. As such, the `layout` method probably should split up into a `layoutSync` and `layoutAsync` based on the setting.

As a side note: I would never expect `layout` to return `Dimension | undefined`, if we return early from the method, I would suggest to return the default fixed size as if wrapping was disabled.

I looked at your changes and I think the direction is good. I think though the max dimensions is not 100% accurate. If you use `editorGroupView#minimumHeight`, I think this would not be the editor control minimum height, but rather the entire minimum height. What we want here is `this.editorControl.minimumHeight`.

SneakyFish5 commented:
Laying out multi-line tabs synchronously seems to have fixed a lot of issues.

> As a side note: I would never expect layout to return Dimension | undefined, if we return early from the method, I would suggest to return the default fixed size as if wrapping was disabled.

Returning the default height is no issue but then what should be returned as the width in the Dimension, since the `dimension` parameter is possibly undefined as well?

>What we want here is this.editorControl.minimumHeight.

Done

bpasero commented:
> Returning the default height is no issue but then what should be returned as the width in the Dimension, since the dimension parameter is possibly undefined as well?

Yeah true, that case I forgot about.

SneakyFish5 commented:
Were there any other issues to be taken care of?

bpasero commented:
I might need some time before I can do another review, then I can tell.

bpasero commented:
@SneakyFish5 I pushed another refactoring in https://github.com/microsoft/vscode/pull/106448/commits/9147044deab259a90be70d6247e665c2eb8efc68, trying to explain here:
* I removed [`editorHeight`](https://github.com/microsoft/vscode/pull/106448/commits/9147044deab259a90be70d6247e665c2eb8efc68#diff-7d793052e71b84617d8dffd73fd161fc4798523d911c223b0f8a3eb20e3f5832L140) which was not used anymore
* I added a `ITitleControlDimensions` to have one property for both the available width/height as well as the maximum
* I make sure that `this.dimension` in `tabsTitleControl` is never `undefined`, even when created. this was causing a lot of weird checks that are now nicer
* `ITitleControlDimensions` is now passed all the way down to `doLayoutTabs` and instead of having a field `disableMultiLineTabs`, it is now decided down there wether multiline can be enabled or not based on available height
* I removed the code to [`group.relayout`](https://github.com/microsoft/vscode/pull/106448/commits/9147044deab259a90be70d6247e665c2eb8efc68#diff-2cc308dd268844b603ed595ef8cbb6caac75a72a42d51024185932c6df0ac719L1033), this needs a different solution (see my last comment here)
* I [touched on the algorithm](https://github.com/microsoft/vscode/pull/106448/commits/9147044deab259a90be70d6247e665c2eb8efc68#diff-2cc308dd268844b603ed595ef8cbb6caac75a72a42d51024185932c6df0ac719R1386) to decide wether to enable or disable multiline tabs, but I am not 100% sure it is still working as you designed it, so please review

Finally, I think the previous `group.relayout` needs a solution. Previously you would only call this from the `redraw` method but truth is that any call to `doLayoutTabs` has a chance of either enabling or disabling multi line tabs, so we really need to make sure that `group.relayout` is triggered whenever the height of the tabs change. I think it would make sense to store the last computed height of the tabs control whenever layout is triggered and then do a comparison of the new height to the old height. If we find out that the height changed, this needs to propagate to the editor control.

PS: I might have to throw the towel on this one for October, I cannot spend more time on this atm.

SneakyFish5 commented:
I intend to fully look into this `group.relayout` this coming weekend as well as the refactoring you’ve done, got a bit busy recently but I should have time to complete the requested changes.

> PS: I might have to throw the towel on this one for October, I cannot spend more time on this atm.

Of course, no worries there. This PR has gone through a multitude of refactoring and changes and I’m sure there are other, more pressing issues. I’ll try to get my fixes out this week/weekend but feel free to review it when you are able to do so 👍.

SneakyFish5 commented:
Pushed

SneakyFish5 commented:
Just pushed https://github.com/microsoft/vscode/pull/106448/commits/2be28a88416498f160ddf8b44bea3b2863ce0a5a with a very simple change that I hope implements what you were looking for? As of now it just stores the current height in `lastComputedHeight` right from the start, and then on every call to layout it checks if the height has changed, and if it has then it calls `group.relayout`. This seems to correctly update the editor size, however there is one problem I noticed.

I'm not sure when this bug was introduced, but whenever you open a tab with breadcrumbs enabled, it permanently breaks multiline tabs and causes the title size to take up the whole container whenever the `multi-line` class is removed (when number of rows is back to 1). That explanation may not have been the best so if you need a recording, I could also provide that tomorrow. I couldn't quite narrow down the cause of this issue, so I was wondering if maybe you might know what it is?

Again, feel free to review whenever you are able to, thanks!

bpasero commented:
Somehow we are always adjusting the height of the tab scrollbar based on the height of the title control, but I think that is not needed:

https://github.com/microsoft/vscode/blob/962bedc4f47b26e4a6ddc917ee2a4d9445531fb9/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1288

I pushed a fix and will also try to push these changes to `master` to reduce the diff of this PR to the minimum.

SneakyFish5 commented:
Awesome, breadcrumbs now works for me. Since you are going to push these other changes you mentioned to `master`, is there anything else that needs to be accomplished/fixed? Either way I'll make sure to do some more testing.

bpasero commented:
@SneakyFish5 I pushed some changes but left the overall behaviour pretty much as you had it. Some changes worth mentioning:
* I removed 2 CSS classes that I felt were not needed, specifically for the `height: auto !important` for the scrollbar (which I think is not needed but maybe you recall why it was added) and I feel that the `:after` element was also not needed and I rather want taps to simply fill the width (but this is something I was planning to get more UX feedback from the team)
* I fixed an issue where wrapping would not properly turn off when the minimum editor height is reached
* I renamed the setting to `workbench.editor.experimentalWrapTabs` because I want to give insider users a chance to test this and report feedback but I am considering to pull it out if there is bad regressions that need more time to fix

I would appreciate if you could give this a good test:
* corner cases such as tabs stop to wrap when exceeding available height
* editor height is always proper, either when changing window size or when size changes from tab modifications within (e.g. closing a tab)
* breadcrumbs on/off
* tab drag and drop
* tab related settings:
  * `tabCloseButton: left | right | off`
  * `tabSizing: fit | shrink`
  * ...

**Next steps:**
* present to team and UX call (today)
* polish based on feedback
* push to insiders to get more feedback behind experimental setting

bpasero commented:
Here is the result of the **UX meeting** with some additional requirements before we can merge this:
* the editor actions on the top right should move to bottom right in the last row and should not consume the entire vertical space so that the rows above can fill the entire width
* when `workbench.editor.tabSizing: shrink` is configured, tabs should not fill the last row and all tabs should have an even size across all rows
* we should have a limit as to how many rows we support (e.g. 3 or 4) and not just rely on the minimal editor height

I will comment in https://github.com/microsoft/vscode/issues/70413 about the status of this PR.

SneakyFish5 commented:
Sorry got a bit tied up but I can now dedicate some more time to this PR. I'm glad things are moving along nicely, and I'll get to work right away on testing those things you listed + more and report back. As for the results of the UX team, there was a couple things I wanted to go over.


> the editor actions on the top right should move to bottom right in the last row and should not consume the entire vertical space so that the rows above can fill the entire width

Might need some help with this one, since I'm assuming it requires some CSS changes? I was experimenting with `align-self: flex-end;` on the `editor-actions` class, but it seems that the `tabsContainer` is separate from the `tabsAndActionsContainer` that the editor actions are in, so that still leaves space. Overall, I'll probably need help with this one 😅, so feel free to chime in CSS experts.

> when workbench.editor.tabSizing: shrink is configured, tabs should not fill the last row and all tabs should have an even size across all rows

Unless I'm misunderstanding, I think this goes back to the issue I was having, how should we make the tabs NOT fill the whole row? I couldn't figure out the CSS to do it, maybe someone else can help me out here? I'll continue to look into though of course.

> we should have a limit as to how many rows we support (e.g. 3 or 4) and not just rely on the minimal editor height

Just pushed a commit for this, it simply replaces the existing logic we had for limiting based on editor minimum height for a new `TabsTitleControl.MAX_WRAPPED_HEIGHT` variable. Currently it's set to 4 rows, can be easily changed, or can add a setting to configure as well? Any changes that need to be made here just let me know.

bpasero commented on src/vs/workbench/browser/parts/editor/tabsTitleControl.ts:
```diff
@@ -1437,13 +1438,14 @@ export class TabsTitleControl extends TitleControl {
-			// Tabs wrap multiline: remove wrapping if height exceeds available height
-			if (tabsWrapMultiLine && tabsContainer.offsetHeight > dimensions.available.height) {
+			// Tabs wrap multiline: remove wrapping if height exceeds the maximum
+			if (tabsWrapMultiLine && tabsContainer.offsetHeight > TabsTitleControl.MAX_WRAPPED_HEIGHT) {
```
@SneakyFish5 shouldn't we check on both `dimensions.available.height` and `MAX_WRAPPED_HEIGHT`? What if the available height is less than `MAX_WRAPPED_HEIGHT`?

SneakyFish5 replied:
Oh right good point, I re-added that in, thanks. Changed max amount of rows to 3 as well as requested.

bpasero commented:
> Sorry got a bit tied up but I can now dedicate some more time to this PR. I'm glad things are moving along nicely, and I'll get to work right away on testing those things you listed + more and report back. As for the results of the UX team, there was a couple things I wanted to go over.

Thanks 👍 

> Might need some help with this one, since I'm assuming it requires some CSS changes? I was experimenting with align-self: flex-end; on the editor-actions class, but it seems that the tabsContainer is separate from the tabsAndActionsContainer that the editor actions are in, so that still leaves space. Overall, I'll probably need help with this one 😅, so feel free to chime in CSS experts.

If it helps to move the tab actions container somewhere else, I am also fine with that provided it does not break anything else.

> Unless I'm misunderstanding, I think this goes back to the issue I was having, how should we make the tabs NOT fill the whole row? I couldn't figure out the CSS to do it, maybe someone else can help me out here? I'll continue to look into though of course.

Isn't this as easy as not setting `flex-grow: 1`? That is the thing that causes tabs to grow when wrapped, no?

> Just pushed a commit for this, it simply replaces the existing logic we had for limiting based on editor minimum height for a new TabsTitleControl.MAX_WRAPPED_HEIGHT variable. Currently it's set to 4 rows, can be easily changed, or can add a setting to configure as well? Any changes that need to be made here just let me know.

Thanks, I commented. I think hardcoding it for now is fine, maybe we should use 3 instead of 4.

SneakyFish5 commented:
> If it helps to move the tab actions container somewhere else, I am also fine with that provided it does not break anything else.

Alright I'll give it a shot, still have to learn how the editor actions work but I'll try my best.

> Isn't this as easy as not setting flex-grow: 1? That is the thing that causes tabs to grow when wrapped, no?

Well here's what setting `tabSizing: shrink` currently looks like:

![ZQQBOEo90o](https://user-images.githubusercontent.com/32284796/102306181-3cb98e00-3f27-11eb-9f0f-c07d01249882.gif)


Not familiar with how `tabSizing: shrink` works, but when I tested it with and without `flex-grow`, it seemed to act the exact same.  Does this fill the requirement of tabs having an even sizing across rows, or am I missing something?

bpasero commented:
Yes I think `tabSizing: shrink` looks good to me, but I wonder why we then still need the `flex-grow: 1;` CSS rule? Shouldn't it only apply when `tabSizing: fit`?

SneakyFish5 commented:
Yea I agree, I switched the `wrap` CSS class to only enable for when `tabSizing: fit` is selected.

bpasero commented:
There is a bug when wrapping is enabled and the editor group is small enough that we switch to non-wrapping mode that I cannot explain:

![recording](https://user-images.githubusercontent.com/900690/102457034-63201c00-4042-11eb-8b22-08eedb890489.gif)

We always try to keep the currently active tab visible via some logic in the `doLayout` code:

https://github.com/microsoft/vscode/blob/3229991032ca6fb0aeeb9e3afeb98db98fdb9fb3/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1420-L1420

For some reason this logic breaks due to us first adding the `wrap` class and then removing it again when we figure out the width is not sufficient. I have not been able to fix this, would appreciate a second pair of eyes looking at it.

SneakyFish5 commented:
> There is a bug when wrapping is enabled and the editor group is small enough that we switch to non-wrapping mode that I cannot explain

Just looked into this and I think I might know where the problem is coming from. The way I limited the rows was calculating the max height in pixels. So `TabsTitleControl.MAX_WRAPPED_HEIGHT` = `35 x 3 = 105px`. So when the editor group is small, that space is quickly filled up causing wrapped lines to be removed. If you try removing some of the tabs in that small editor group, does it begin to work for you? Or were you talking about a different issue?

> We always try to keep the currently active tab visible via some logic in the doLayout code

Yup I see what you're talking about, I'll try digging into it to see what I can find.

bpasero commented:
@SneakyFish5 figured it out: when wrapping is disabled, we need to give the scrollbar an update so that the logic for revealing can function properly: https://github.com/microsoft/vscode/pull/106448/commits/fe75f6cfd8a61ad68c2deb6e8a449551005d8f60

bpasero commented:
Oh, I was able to drop the sync layout altogether given the `relayout` logic (via https://github.com/microsoft/vscode/pull/106448/commits/e69e7c31c503a902fa6a0f85db6753687351c08a). This is cool because it means when tabs wrapping is enabled, there should not be a bad perf impact for clients. 

I think we are down to just 1 remaining task here: move the editor actions to the bottom right 👍 . And then do extensive testing as I outlined in https://github.com/microsoft/vscode/pull/106448#issuecomment-741666411

DynDux commented:
I'm not a programmer but deal with css a lot. Maybe I have a solution for the placing of the editor actions.
Please have a look at the fiddle I created: https://jsfiddle.net/t1fpwnxk/

It works if you are able to identify the tabs of the last row and give them a "flex: 0"

I hope it's of use for you because I'm eager to get tabs wrapping in VSCode. :-)
Thanks for this great tool!

SneakyFish5 commented:
@bpasero that's awesome, glad the perf impact will be reduced.

I've gone ahead and implemented part of the CSS changes that you've brought forward @DynDux. Thank you so much for your help, here's a recording of what I've got so far.

![SvZDHsn5En](https://user-images.githubusercontent.com/32284796/102657451-4d425200-413b-11eb-993b-b8e83ecd8e71.gif)

Pretty much what I've done is add a new `wrap` class to the editor actions, and add/remove it with the same logic that we add/remove the `wrap` class from the tabs container. If there is a better way to accomplish this I'm all for it. As for the class itself, I added the properties that you have in your fiddle for the icons/editor actions, but as you can see in the gif, they're sort of on top of the tabs. Is there any other property/CSS changes that can fix that, or maybe something I missed adding from the fiddle that would fix this issue?

DynDux commented:
I'm happy to be able to assist!
According from what I see from your screen recording, two things are missing:

1) the last tab has to have a margin-right of the width of the editor actions - maybe with some more pixels padding so it doesn't stick to the actions (css: line 44 in fiddle - sorry, forgot to write what width it should be..):

`.tabs div:last-child {
    margin-right: 150px; /* width of editor actions + some padding */
}`

2) the tabs **only on the last row** have to get a `flex: 0` to not span the full width of the container. I gave some tabs the class "lastrow" in fiddle but of course it depends on what tabs are really in the last row. Don't know if/how that's possible to accomplish but this is essential to not possibly get very large tabs on the last row (css: line 63 in fiddle):

`.fillwidth .tabs div.lastrow {
  flex: 0; /* disable fill only on lastrow tabs */
}`

SneakyFish5 commented:
That is completely my fault, missed that part of the fiddle. Here's an updated recording of what I've got so far (I also pushed the commit, any feedback is much appreciated).

![v3sw8vwRRu](https://user-images.githubusercontent.com/32284796/102708437-bbd0fe00-4268-11eb-9138-94db5c9b94de.gif)

Works with breadcrumbs as well. I tried to explain the logic with comments as best as I could, but I can definitely try to elaborate if need be.  There is however one issue that I can't seem to figure out how to fix. It happens whenever the `wrap` class is removed and then applied again (when the title height exceeds the 3 row limit). Here's a visual example, since I can't quite describe it:

![dNGLqJvcCi](https://user-images.githubusercontent.com/32284796/102708511-66e1b780-4269-11eb-9057-14aabbf11d97.gif)

@bpasero any chance you know a potential fix to this issue? It doesn't happen all the time, only when the window is certain sizes. Aside from that, does this look like what the UX team was looking for?

Again thank you @DynDux, you're a real lifesaver :smile:.

DynDux commented:
@SneakyFish5 I can see you implemented the margin-right on the last tab but from what I see from the screen recording, the style for the _last row_ is missing (point 2 of my last commit). The difference is the following:

**- with no extra style applied (flex: 0)**
![tabs_flex_1](https://user-images.githubusercontent.com/76206890/102714583-d38f9d00-42cf-11eb-89f4-90c1a476b420.png)

**- with added extra styles (flex: 1):**
![tabs_flex_0](https://user-images.githubusercontent.com/76206890/102714567-ac38d000-42cf-11eb-8273-33447c3144bb.png)

But, again, I don't know if it's possible to know, what tabs are on the last row and it has to change with each window resizing or opening/closing files...

I also think it looks nicer, if you add a 1px margin to the bottom of the tabs so they have a visual spacing (look close!):
![tabs_margin-bottom](https://user-images.githubusercontent.com/76206890/102714615-28cbae80-42d0-11eb-9ce0-a4bcbeaed87d.png)

If I may add one more request or suggestion to the row count on which the tabs revert to single line: I sometimes deal with 5+ rows of tabs (with custom css in vscode) because of my workflow (I know... ;-)) and I feel switching from multiline to single line would be irritating when opening or closing just one file. It would be great to choose on how many rows this happens or disable it altogether.
The number of rows also greatly depends on the width of the window and the length of the filenames - so when coding on a notebook, you might only have a few files open.

Maybe the better solution for small screens would be to scroll the tabs vertically instead of reverting to single line?

SneakyFish5 commented:
Again, my fault should've mentioned that. I thought per https://github.com/microsoft/vscode/pull/106448#issuecomment-742371268, the tab sizing on the last row was only for `tabSizing: shrink`, which was apparently already solved? Either way, I think that was separate from the editor actions thing, unless I'm mistaken?

> I also think it looks nicer, if you add a 1px margin to the bottom of the tabs so they have a visual spacing (look close!):

Yea this was part of the very first commit, however got taken out for some reason that I can't quite remember now (theming issues?). I think @bpasero was contemplating adding it back, and I would be happy to take that up.

As for the 3 row limit and scrollbar, I personally am fine either way so I'll leave that final decision up to what the community, @bpasero, and the VSCode team as a whole decide.

DynDux commented:
Thanks for your work and starting this PR! I'm sure the team and community will decide what is best. :-)

bpasero commented:
@SneakyFish5 this looks great and what I had in mind with regards to the actions aligning to the right of the last row 👍 

> There is however one issue that I can't seem to figure out how to fix. 

Can you check if this was caused by any of the new changes to rule out a regression from these?

> Again, my fault should've mentioned that. I thought per #106448 (comment), the tab sizing on the last row was only for tabSizing: shrink, which was apparently already solved? Either way, I think that was separate from the editor actions thing, unless I'm mistaken?

Now that we have the actions on the last row and the tabs above filling the entire row, I wonder if we should revisit the `flex-grow: 1` for tabs when wrapping. Maybe we leave it up to the user how to configure the `tabSizing` to drive this and not interfere here. Happy to try this out though once this PR is in a state to play with it again.

> I also think it looks nicer, if you add a 1px margin to the bottom of the tabs so they have a visual spacing (look close!):

I like the border and had thought about it. Would this simply be a "gap" to the background or themable? Ideally a user would be able to change it's color but it might be challenging to accomplish given all the other border colors you can set per tab.

> As for the 3 row limit and scrollbar, I personally am fine either way so I'll leave that final decision up to what the community, @bpasero, and the VSCode team as a whole decide.

This sounds like a setting we need to introduce. But I am also fine to gather some feedback from actual users of this feature before making a decision. The reason I thought this was a good idea is that it could get very frustrating if the size of the editor is constantly limited by the fact that tabs above wrap and take valuable space away. This can happen quickly when you make one editor group very large and the other very small.

SneakyFish5 commented:
> Can you check if this was caused by any of the new changes to rule out a regression from these?

Oh sorry should have been more clear. It is due to the new changes, and I believe specifically due to these two lines: https://github.com/microsoft/vscode/blob/4585b7042fbd87aa461c3de8201df537e54bbc09/src/vs/workbench/browser/parts/editor/tabsTitleControl.ts#L1449-L1450

Was wondering if maybe you knew a reason for why it has that jarring effect, or possibly a tip on solving it?

> I wonder if we should revisit the flex-grow: 1 for tabs when wrapping

Sure, I can definitely start looking into that. To be clear, keep `tabSizing: shrink` how it currently is, but work on making `tabSizing: fit` not grow the tabs with `flex-grow: 1`. Rather they keep their default "fitting" width until the row doesn't have enough space for a new tab, then grow them (to fill the empty space at the end of the row)?

> I like the border and had thought about it. Would this simply be a "gap" to the background or themable?

I was thinking themable, and just using the same `TAB_BORDER` that is used for the right border on the tabs. What do you think?

bpasero commented:
Fyi I will be out starting today and maybe slow to respond, I suggest to get this into a bug-free state and then we can think about additional features such as the border or how tabs should grow 👍

bpasero commented:
@SneakyFish5 fyi pushed https://github.com/microsoft/vscode/pull/106448/commits/b4c809de8090882e9f9ed2363420676b3d0beeec with the following impact:
* removed the `3` row limit for wrapping because I think this probably needs a setting if enforced and can always come later if a majority of people ask for it (maybe you can check if that fixes your issue too)
* moved the `wrapping` class to the parent `tabs-and-actions-container` so that we don't have to add this class to both tabs and actions
* changed the `tabs-and-actions-container` to be `position: relative` so that the actions can be positioned fully through CSS and don't have to be updated depending on breadcrumbs to be visible or not

I am still not entirely happy with the `margin-right` code (`tabsContainer.lastChild?.previousSibling....`), though it works reasonably well. An alternate idea I had was to treat the actions container like a tab that would simply always be added as last tab and flow with the others at the end, but that might require even more changes....

**PS:** I played again with `fit` vs `shrink` for the tab sizing setting when wrapping is on and feel that we can leave the `flex-grow: 1` for now when `fit`, I think this works reasonably well even if it means that tabs on the last row grow artificially.

bpasero commented:
Actually with https://github.com/microsoft/vscode/pull/106448/commits/a53fe3c3414f065a423a7607dcaa70b22d5b5861 found a nice solution via CSS variables that does not require to patch any tab element directly. This makes sure we only have a `margin-right` for the last tab when wrapping is enabled.

I think this PR is close to be done, I would appreciate some [testing](https://github.com/microsoft/vscode/pull/106448#issuecomment-741666411) so that we are certain this works as intended.

cullenjohnson commented:
This is the PR I have been waiting for for years. Is there any way I can help get this merged faster? :)

SneakyFish5 commented:
Hey, sorry again I've been very busy recently. Thank you so much @bpasero for pushing all those fixes.

@cullenjohnson maybe you can help with the testing outlined [here](https://github.com/microsoft/vscode/pull/106448#issuecomment-741666411), or just in general do any sort of testing that you can think of? If not no worries, I'll be able to get some testing done within the next day or so as well, but of course all help is appreciated.

swinder0161 commented:
Hi,

I am up for testing but where will I get a version to test. Donot have
development environment to build myself.

I need a Mac version btw.

Thanks,
Swinder

On Wed, Dec 30, 2020, 12:51 PM SneakyFish5 <notifications@github.com> wrote:

> Hey, sorry again I've been very busy recently. Thank you so much @bpasero
> <https://github.com/bpasero> for pushing all those fixes.
>
> @cullenjohnson <https://github.com/cullenjohnson> maybe you can help with
> the testing outlined here
> <https://github.com/microsoft/vscode/pull/106448#issuecomment-741666411>,
> or just in general do any sort of testing that you can think of? If not no
> worries, I'll be able to get some testing done within the next day or so as
> well, but of course all help is appreciated.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/microsoft/vscode/pull/106448#issuecomment-752357262>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADBBZ3QINBMPYNUEI2DSZDSXLIIFANCNFSM4RGQAQ6A>
> .
>

SneakyFish5 commented:
Sorry for the delay again, had a lot of issues with `node-gyp` when building but was finally able to get it working and do the tests. I did a run through of all things outlined in https://github.com/microsoft/vscode/pull/106448#issuecomment-741666411 as well as:

- Anything I could think of related to pinned tabs
- `tabCloseButton` setting
- `titleScrollbarSizing` setting
- Scroll to switch tabs
- Theming

I also tested a few more things that I can't quite remember now, however they all worked, and everything looked good in regards to the editor height and layout. If there are any more tests that need to be completed, please let me know.

bpasero commented:
Thanks a lot @SneakyFish5 I plan to merge this soon to get it into insiders during January for more coverage.

bpasero commented:
Thanks @SneakyFish5 for pushing this forward 🥂

Tschebbe commented:
This issue seems to be merged now - Am I right?
When is the next official release of VSCode?

bpasero commented:
@Tschebbe if you don't want to wait until your next stable release in early February, simply switch to VSCode insiders to get this as early as tomorrow: https://code.visualstudio.com/insiders/

SneakyFish5 commented:
Thank you as well @bpasero for your continuous effort and assistance in pushing this PR forward!

DynDux commented:
I think I found a pure css solution for the problem with last tab spanning the whole width!
Just add the following style when _tabSizing = fit_:

`.tabs-container::after{
  content: '';
  flex-grow: 20;
}`

This will create an additional tab pseudo-element, that grows 20 times larger than the other tabs. It could be an even higher number, but I think 20 should be sufficient.
The `margin-right` on the last-child ist still needed though.

bpasero commented:
@DynDux @SneakyFish5 great, I suggest we take this idea to a new PR and test it out there, what do you think?

DynDux commented:
Totally ok for me! Just thought it belongs to this PR because this is how it should have been resolved in first place. ;-)
Please do whatever the right workflow is. Thank you so much!

bpasero commented:
I have added your suggested solution to https://github.com/microsoft/vscode/issues/113801.

qs-labs commented:
Great work to those involved! Pumped to have this feature 👏 🎉 👏
